common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,351 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/*
+Package algorithm contains a running-weighted-average calculator for ratelimits,
+and some associated types to prevent accidental confusion between the various
+floats/ints/etc involved.
+
+---
+
+This package is intentionally unaware of RPC or any desired limits, it just tracks
+the observed rates of requests and returns proportions, so other code can
+determine RPS values.
+
+This is both to simplify testing and to keep its internals as fast as possible, as
+it is relatively CPU heavy and needs to be quick to prevent a snowballing backlog
+of concurrent updates and requests for aggregated data.
+*/
+package algorithm
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/constraints"
+
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/dynamicconfig"
+)
+
+type (
+	// history holds the running per-second running average for a single key from a single host, and the last time it was updated.
+	history struct {
+		lastUpdate         time.Time // only so we know if elapsed times are exceeded, not used to compute per-second rates
+		accepted, rejected PerSecond // requests received, per second (conceptually divided by update rate)
+	}
+
+	// Identity is an arbitrary (stable) identifier for a "limiting" host.
+	Identity string
+	// Limit is the key being used to Limit requests.
+	Limit string
+	// PerSecond represents already-scaled-to-per-second RPS values
+	PerSecond float64
+	// HostWeight is a float between 0 and 1, showing how much of a ratelimit should be allocated to a host.
+	HostWeight float64
+
+	impl struct {
+		cfg Config
+
+		// mut protects usage, as it is the only mutable data
+		mut sync.Mutex
+		// usage data for ratelimits, per host.
+		//
+		// data is first keyed on the limit and then on the host, as it's assumed that there will be
+		// many times more limit keys than limiting hosts, and this will reduce cardinality more rapidly.
```

Comment:
> many times more limit keys than limiting hosts
Currently it's the opposite. Did you think of the workflowid based rate limits?

Reply:
I was asking because without workflowid based rate limits in the picture I thought we would have more hosts than limit keys. Limits in the order of endpoint groups (internal, external etc.) was the assumption but just realized it should be in the order of domains * endpoint_groups. So ignore this comment

Reply:
aah, gotcha.  yeah, this is literally a "per [finest-grained thing being limited]" key (1:1 with the `golang.org/x/time/rate.Limiter` instances), so we'll start with several hundred or thousands in basically all of our clusters, and might go a couple orders of magnitude larger eventually.